awx_collection/plugins/module_utils/tower_api.py

```diff
@@ -105,32 +105,36 @@ def get_one(self, endpoint, *args, **kwargs):
 
         return response['json']['results'][0]
 
-    def resolve_name_to_id(self, endpoint, name_or_id):
-        # Try to resolve the object by name
+    def get_one_by_name_or_id(self, endpoint, name_or_id):
         name_field = 'name'
         if endpoint == 'users':
             name_field = 'username'
 
-        response = self.get_endpoint(endpoint, **{'data': {name_field: name_or_id}})
-        if response['status_code'] == 400:
-            self.fail_json(msg="Unable to try and resolve {0} for {1} : {2}".format(endpoint, name_or_id, response['json']['detail']))
+        query_params = {'or__{0}'.format(name_field): name_or_id}
+        try:
+            query_params['or__id'] = int(name_or_id)
+        except ValueError:
+            # If we got a value error than we didn't have an integer so we can just pass and fall down to the fail
+            pass
+
+        response = self.get_endpoint(endpoint, **{'data': query_params})
+        if response['status_code'] != 200:
+            self.fail_json(msg="Failed to query endpoint {0} for {1} {2} ({3}), see results".format(endpoint, name_field, name_or_id, response['status_code']), resuls=response)
 
         if response['json']['count'] == 1:
-            return response['json']['results'][0]['id']
+            return response['json']['results'][0]
+        elif response['json']['count'] == 2:
+            for tower_object in response['json']['results']:
+                if tower_object['name'] == name_or_id:
+                    return tower_object
```

Comment:
What scenario is this `elif` trying to address?

If it's addressing the case of >1 object with the same name, then that stands in disagreement to the error case below:

> Found too many names {0} at endpoint

If there are >1 objects with the same name, it should hit that error, and "guessing", and returning the first object here is contradictory.

----

If, on the other hand, this block is designed to fish out an object with "id" equal to `name_or_id` when there are _other_ objects with the integer as a string "name"... then other details are not right. In that scenario, this should not be only for `== 2` objects, you should run this when the count is anything `>1`. You should also only exit when `tower_object['id'] == name_or_id`, as opposed to 'name'.